spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -0,0 +1,371 @@
+
+      val alterTableSql =
+        s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN value TYPE ${testCase.toType.sql}"
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(alterTableSql)
+        },
+        errorClass = "NOT_SUPPORTED_CHANGE_COLUMN",
+        sqlState = None,
+        parameters = Map(
+          "table" -> s"`spark_catalog`.`delta`.`$tempPath`",
+          "originName" -> toSQLId("value"),
+          "originType" -> toSQLType(testCase.fromType),
+          "newName" -> toSQLId("value"),
+          "newType" -> toSQLType(testCase.toType)),
+        context = ExpectedContext(
+          fragment = alterTableSql,
+          start = 0,
+          stop = alterTableSql.length - 1)
+      )
+    }
+  }
+
+  test("type widening in nested fields") {
+    sql(s"CREATE TABLE delta.`$tempPath` " +
+      "(s struct<a: byte>, m map<byte, short>, a array<short>) USING DELTA")
+    append(Seq((1, 2, 3, 4))
+      .toDF("a", "b", "c", "d")
+      .selectExpr(
+        "named_struct('a', cast(a as byte)) as s",
+        "map(cast(b as byte), cast(c as short)) as m",
+        "array(cast(d as short)) as a"))
+
+    assert(readDeltaTable(tempPath).schema === new StructType()
+      .add("s", new StructType().add("a", ByteType))
+      .add("m", MapType(ByteType, ShortType))
+      .add("a", ArrayType(ShortType)))
+
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN s.a TYPE short")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN m.key TYPE int")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN m.value TYPE int")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN a.element TYPE int")
+
+    assert(readDeltaTable(tempPath).schema === new StructType()
+      .add("s", new StructType().add("a", ShortType))
+      .add("m", MapType(IntegerType, IntegerType))
+      .add("a", ArrayType(IntegerType)))
+
+    append(Seq((5, 6, 7, 8))
+      .toDF("a", "b", "c", "d")
+        .selectExpr("named_struct('a', cast(a as short)) as s", "map(b, c) as m", "array(d) as a"))
+
+    checkAnswer(
+      readDeltaTable(tempPath),
+      Seq((1, 2, 3, 4), (5, 6, 7, 8))
+        .toDF("a", "b", "c", "d")
+        .selectExpr("named_struct('a', cast(a as short)) as s", "map(b, c) as m", "array(d) as a"))
+  }
+
+  test("type widening using ALTER TABLE REPLACE COLUMNS") {
+    append(Seq(1, 2).toDF("value").select($"value".cast(ShortType)))
+    assert(readDeltaTable(tempPath).schema === new StructType().add("value", ShortType))
+    sql(s"ALTER TABLE delta.`$tempPath` REPLACE COLUMNS (value INT)")
+    assert(readDeltaTable(tempPath).schema === new StructType().add("value", IntegerType))
+    checkAnswer(readDeltaTable(tempPath), Seq(Row(1), Row(2)))
+    append(Seq(3, 4).toDF("value"))
+    checkAnswer(readDeltaTable(tempPath), Seq(Row(1), Row(2), Row(3), Row(4)))
+  }
+
+  test("row group skipping Short -> Int") {
+    withSQLConf(
+      SQLConf.FILES_MAX_PARTITION_BYTES.key -> 1024.toString) {
```

Comment:
How do we verify that this actually leads to row group skipping? FWIW it looks like this should lead to regular min/max data skipping on the Delta log. The max partition bytes only affects how we read chunks, i.e., it'll split a parquet file of >1024 bytes over multiple tasks (1024 bytes each), but if the file is still a single row group, then only one of the tasks will actually read it (the task that gets the "middle" byte of the row group), and the other files will simply ignore the row group. And with the current way of writing, each file will have only one row group, so the row group skipping will be equal to the Delta-level file skipping, so it should never skip, really.

Reply:
This test doesn't make a lot of sense here anymore, I added it back when I found the overflow issue in parquet row group skipping and kept it around but it's redundant with the test I added when I fixed the issue in Spark: 
https://github.com/databricks/runtime/blob/a9d6a9d6191264ced1f0658ff6675b0f30e8e77f/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala#L1216

I'm removing it, I don't think it's worth the effort to get it to accurately cover an issue that the Spark test already covers a lot better